ICSharpCode.Decompiler/TypeSystem/MetadataModule.cs

```diff
@@ -477,77 +477,95 @@ IMethod ResolveMethodReference(MemberReferenceHandle memberRefHandle, GenericCon
 			{
 				throw new BadImageFormatException($"Member reference must be method, but was: {memberRef.GetKind()}");
 			}
-			MethodSignature<IType> signature;
+			MethodSignature<IType> signature = default;
 			IReadOnlyList<IType> classTypeArguments = null;
-			IMethod method;
+			IMethod method = null;
 			if (memberRef.Parent.Kind == HandleKind.MethodDefinition)
 			{
 				method = ResolveMethodDefinition((MethodDefinitionHandle)memberRef.Parent, expandVarArgs: false);
 				signature = memberRef.DecodeMethodSignature(TypeProvider, context);
 			}
 			else
 			{
-				var declaringType = ResolveDeclaringType(memberRef.Parent, context);
-				var declaringTypeDefinition = declaringType.GetDefinition();
-				if (declaringType.TypeArguments.Count > 0)
-				{
-					classTypeArguments = declaringType.TypeArguments;
-				}
-				// Note: declaringType might be parameterized, but the signature is for the original method definition.
-				// We'll have to search the member directly on declaringTypeDefinition.
 				string name = metadata.GetString(memberRef.Name);
-				signature = memberRef.DecodeMethodSignature(TypeProvider,
-					new GenericContext(declaringTypeDefinition?.TypeParameters));
-				if (declaringTypeDefinition != null)
+				var statedDeclaringType = ResolveDeclaringType(memberRef.Parent, context);
+				var statedDeclaringTypeDefinition = statedDeclaringType.GetDefinition();
+
+				if (statedDeclaringTypeDefinition != null)
 				{
-					// Find the set of overloads to search:
-					IEnumerable<IMethod> methods;
-					if (name == ".ctor")
-					{
-						methods = declaringTypeDefinition.GetConstructors();
-					}
-					else if (name == ".cctor")
-					{
-						methods = declaringTypeDefinition.Methods.Where(m => m.IsConstructor && m.IsStatic);
-					}
-					else
-					{
-						methods = declaringTypeDefinition.GetMethods(m => m.Name == name, GetMemberOptions.IgnoreInheritedMembers)
-							.Concat(declaringTypeDefinition.GetAccessors(m => m.Name == name, GetMemberOptions.IgnoreInheritedMembers));
-					}
-					// Determine the expected parameters from the signature:
-					ImmutableArray<IType> parameterTypes;
-					if (signature.Header.CallingConvention == SignatureCallingConvention.VarArgs)
-					{
-						parameterTypes = signature.ParameterTypes
-							.Take(signature.RequiredParameterCount)
-							.Concat(new[] { SpecialType.ArgList })
-							.ToImmutableArray();
-					}
-					else
+					var candidateDeclaringTypes = new List<IType>() { statedDeclaringType };
+
+					// Search through type hierarchy of declaring type, traversing up the chain.
+					if (statedDeclaringTypeDefinition.DirectBaseTypes?.Any() == true)
 					{
-						parameterTypes = signature.ParameterTypes;
+						candidateDeclaringTypes.AddRange(statedDeclaringTypeDefinition.DirectBaseTypes);
```

Comment:
This seems strange... why are we only interested in "one" level of base types, instead of the whole hierarchy? The comment above mentions the whole hierarchy.